reafctor(deserialize): geenric config deserialization

Conaclos opened the pull request:
## Summary

This PR is an application of #620.
In contrast to #620, this proposal is closer to the implementation of serde.
In particular, the visitors are dedicated instances that can hold a state. This allows to support more complex use cases such as deserializing the configuration based on a rule name (The rule name is held by the visitor).

The system relies on three main traits defined in `biome_deserialize::visitor`:

- `DeserializationVisitor` which allows visiting a value (array, map, boolean ,string, ...)
- `DeserializableLanguage` that implements the specific code for every language
- `Deserializable` that allows to deserialize a type (this is analog to the `Deserialize` trait of serde).

One advantage of separating `Deserializable` and `DeserializationVisitor` is the possibility of reusing a visitor for several types.

Tasks:

- [ ] Convert the entire code base with the new system
- [ ] use macros for auto-implement `Deserializable` and `DeserializationVisitor`?

## Test Plan

CI should pass.

ematipico commented on crates/biome_deserialize/src/json.rs:
```diff
@@ -613,3 +616,78 @@ where
     }
 }
+
+declare_node_union! {
+    AntJsonEntity =
+        JsonArrayValue
+        | JsonBogusValue
+        | JsonBooleanValue
+        | JsonNullValue
+        | JsonNumberValue
+        | JsonObjectValue
+        | JsonStringValue
+        | JsonMemberName
+}
```
I think we _might_ not need this union anymore. Instead, I think we should receive a `JsonRoot` node instead.

ematipico commented on crates/biome_deserialize/src/json.rs:
```diff
@@ -613,3 +616,78 @@ where
 }
+
+declare_node_union! {
+    AntJsonEntity =
```
`AnyJsonVisitor` ?

ematipico commented on crates/biome_service/tests/invalid/vcs_missing_client.json.snap:
```diff
@@ -15,22 +15,3 @@ vcs_missing_client.json:4:20 deserialize ━━━━━━━━━━━━━
   
 
 
-
```
We lost this diagnostic, which is a regression :/

ematipico commented on crates/biome_js_analyze/src/analyzers/complexity/no_excessive_cognitive_complexity.rs:
```diff
@@ -402,37 +403,50 @@ impl ComplexityOptions {
+    }
+}
+
+impl Display for ComplexityOptionsVisitor {
+    fn fmt(&self, fmt: &mut Formatter) -> std::io::Result<()> {
+        write!(fmt, "object")
```
I am conflicted by the implementation of `Display`, while I understand why it's needed, the reason *why* it needs to be implemented is unclear.

Can't we have a new trait called `ExpectedType` that has a function called `expect_type` that returns a `&'static str` and then pass it to the diagnostic? (instead of passing `self`).

Curious to know what are the constraints

Conaclos replied:
We can same add `expect_type` on the `DeserializationVisitor` trait. Using `Display` has the advantage to be closer to serde and allow custom message based on the visitor fields. Maybe it is overkill since we always use a simple static string...

ematipico replied:
Yeah I think it's an overkill and difficult to justify.

 Also, we are in control of the errors we emit, so we have more freedom

ematipico reviewed (commented):
I like this refactor. I am looking forward to getting this shipped. I left some suggestions and questions, feel free to address them or not. 

Here are some things that we should do (now or after we merge the PR):

- **Documentation**: things now require a different implementation, especially for custom types where we require the usage of some external visitors (Unsize... types for example). Let's craft good documentation with some examples, especially how to deserialize enums, numbers and custom types, e.g. `StringSet`
- Explain when to use `Deserializable` and when to use `Visitor`. Even though we took a lot of inspiration from `serde`, we can't assume that people know how to use it, especially new contributors that want to create rules with options.

Conaclos commented on crates/biome_deserialize/src/json.rs:
```diff
@@ -613,3 +616,81 @@ where
+                    visitor.visit_f64(value, range, diagnostics)
+                } else {
+                    diagnostics.push(DeserializationDiagnostic::new_out_of_bound_number(
+                        i64::MIN,
+                        u64::MAX,
```
There are not. This code path is run when it is not possible to convert a number into u64 and i64 and f64. In other word it cannot be handled by the visitor.

Conaclos commented on crates/biome_js_analyze/src/analyzers/complexity/no_excessive_cognitive_complexity.rs:
```diff
@@ -378,13 +379,13 @@ pub struct ComplexityScore {
     fn default() -> Self {
         Self {
-            max_allowed_complexity: 15,
+            max_allowed_complexity: NonZeroU8::new(15).unwrap(),
```
Unfortunately [new_unchecked](https://doc.rust-lang.org/std/num/struct.NonZeroU8.html#method.new_unchecked) requires an unsafe block. Do you think it is worth it?

ematipico replied:
Oh right. Then keep that way

Conaclos commented on crates/biome_formatter/src/lib.rs:
```diff
@@ -176,6 +179,20 @@ impl Default for LineWidth {
+    ) -> Option<Self> {
+        L::deserialize(
+            value,
+            UnsignedIntegerVisitor::new(Self::MIN, Self::MAX),
```
Because `LineWidth` has a custom maximum (320). This ensures that the value is not greater than 320. SHould we remove this limit? I am not sure why we use 320. That's pretty low?

ematipico replied:
I can't remember unfortunately

Conaclos commented on crates/biome_deserialize/src/json.rs:
```diff
@@ -613,3 +616,81 @@ where
+            }
+            AnyJsonVisitableNode::JsonBooleanValue(value) => {
+                let value = value.value_token().ok()?;
+                visitor.visit_bool(value.kind() == T![true], range, diagnostics)
```
I don't think so. The visitor needs a boolean. If we pass the value, the consumer don't know if it is true or false?

ematipico commented on crates/biome_deserialize/src/visitor.rs:
```diff
@@ -1,39 +1,774 @@
+    }
+}
+
+pub struct IntegerVisitor<T: TryFrom<i64>> {
```
I forgot to add that it's usually idiomatic to add the boundaries only during the `impl` blocks.

We do this throughout the code base.


 I've seen it in a few places. The definition of the struct should be generic over `T`, and the implementation should define the boundary (which is like this already)

Conaclos commented on crates/biome_deserialize/src/diagnostics.rs:
```diff
@@ -54,6 +54,18 @@ impl DeserializationDiagnostic {
+        range: impl AsSpan,
+    ) -> Self {
+        Self::new(markup! {
+            "The number should be between "<Emphasis>{min}</Emphasis>" and "<Emphasis>{max}</Emphasis>"."
```
Hum, I wonder: should we pass a string to `visit_u64`, `visit_i64` and `visit_f64` instead of a numeric type? This could allow arbitrary precision integers / floats.

ematipico replied:
Good call, I think we should

Conaclos commented:
@ematipico
Do you think it could make sense of using a bitflag for `EXPECT_TYPE` instead of a static string? The expected type should always be a combination of the following types: array, map, object, string, integer, unsigned integer, float, boolean?

ematipico commented:
> @ematipico
> Do you think it could make sense of using a bitflag for `EXPECT_TYPE` instead of a static string? The expected type should always be a combination of the following types: array, map, object, string, integer, unsigned integer, float, boolean?

Oh that's clever, I like that.

A small note: object and map are the same from the point of view of a JSON file/user. Let's not use "unsigned integer", but negative numbers - or something similar - instead because it's not a beginner friendly  term.

Conaclos commented:
> A small note: object and map are the same from the point of view of a JSON file/user

Sure. However, a user could think that an object has a fixed number of keys, while a map hasn't.

ematipico commented:
> > A small note: object and map are the same from the point of view of a JSON file/user
> 
> Sure. However, a user could think that an object has a fixed number of keys, while a map hasn't.

Technically, that's true, but a user doesn't need to understand the different.
e.g. `package.json#dependencies` VS `biome.json#formatter`.

Technically, for a **developer** (us), the latter (`biome.json#formatter`) is an object, and the former (`package.json#dependencies`) is a map (their keys aren't typed). But for end users, they are both **objects**.

ematipico commented on crates/biome_deserialize/src/impls.rs:
I think each implementation should have at least a test, either via doc test or normal test.

ematipico commented on crates/biome_deserialize/src/impls.rs:
```diff
@@ -0,0 +1,571 @@
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        let range = value.range();
+        let value = TokenNumber::deserialize(value, diagnostics)?;
```
When I read the documentation of `TokenNumber`, I thought it was meant to deserialize cases like `"49"`, although here we are using it for a `f64`. So, it's possible I don't understand this implementation or the documentation of `TokenNumber` is incorrect.

ematipico commented on crates/biome_deserialize/src/impls.rs:
```diff
@@ -0,0 +1,571 @@
+    }
+}
+
+impl<T: Deserializable + Eq + Hash, S: BuildHasher + Default> Deserializable for HashSet<T, S> {
```
While this is useful, `HashSet` and `Vec` **don't** guarantee insertion order, so we shouldn't provide functionalities for these types. In fact, internally, we use `StringSet` for all the arrays we have, which is a tiny wrapper for `IndexSet`. 

This will bite us if/when we will implement `Serialize`

Conaclos replied:
> [...] Vec don't guarantee insertion order

`Vec` is by design an ordered sequence?

ematipico replied:
Does `Vec` guarantee insertion order? I was under the assumption that it does not.

Conaclos replied:
It depends on how you add items. If you use `push`, data is always appended. Thus, the order is preserved.

ematipico commented on crates/biome_deserialize/src/impls.rs:
```diff
@@ -0,0 +1,571 @@
+            ) -> Option<Self::Output> {
+                Some(
+                    values
+                        .filter_map(|value| Deserializable::deserialize(value, diagnostics))
```
This `filter_map` removes items that contain syntax errors. This means that during the deserialization, the original array and the deserialized array can be different in length. Is this intended? It seems to be very unsafe.

Conaclos replied:
Yes, it is part of the fault-tolerant philosophy. I am not sure to see the problem with that. We also deserialize an object even if one of its property has a syntax error.

ematipico replied:
Cool. Just making sure that that's what we want

ematipico commented on crates/biome_deserialize/src/impls.rs:
```diff
@@ -0,0 +1,571 @@
+    }
+}
+
+impl<K: Ord + Deserializable, V: Deserializable> Deserializable for BTreeMap<K, V> {
```
We should not implement  `HashMap` and `BTreeMap`, both don't retain **insertion** order. We should instead implement `Deserializable` for `IndexMap`

Conaclos replied:
Bijective encoding is an important property to have, however it is more a concern for binary format. I don't see the value of guarantying this property for human-readable formats. We still can't guarantee bijective encoding for objects (and we won't, because a user shouldn't care about the order of object properties). Thus, I don't see the value of bothering for unordered maps and unordered array-like structure. Could you elaborate?

ematipico replied:
Ordering **is important**! Two reasons:
- `biome.json#overrides` are applied in sequence from first to last in the order they are declared
- when we **serialize** `biome.json#overrides`, we want to keep the insertion order

Now, I understand why you want to apply this, and it's fine with me. However, it's also important to explain to our developers when using one type rather than another. For example, our `Configuration` uses only `StringSet` because it requires maintaining insertion order. So, as an action item, I believe we should explain to our contributors which type they should use to represents their data format (most of the time, a JSON).

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+/// When deserializing more complex types, such as `struct`,
+/// you have to use a type that implements [DeserializationVisitor].
+///
+/// ### Example
```
```suggestion
/// ## Example
```

The example is good, although we don't see how the implementation is eventually used, and this is true for the whole file. There isn't an example that shows how a "data format" is deserialized into the data structure defined in any example. 

I believe this is **very important**.

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+///         value: impl DeserializableValue,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self> {
+///         const ALLOWED_VARIANTS: &[&str] = &["A", "B", "C"];
```
```suggestion
///         const ALLOWED_VARIANTS: &[&str] = &["A", "B"];
```

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+/// This trait represents a visitor that walks through a [DeserializableValue].
+///
+/// We assume that a deserializable value has one of the following type:
+/// - null / none
+/// - boolean
+/// - number (integer, floats)
+/// - string
+/// - array
+/// - map (key-value pairs with value's type that can depend on its key)
```
```suggestion
/// - null / none;
/// - boolean;
/// - number - a integer or a float;
/// - string;
/// - array;
/// - map, key-value pairs with value's type that can depend on its key;
```

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+/// Most of the time you should implement [Deserializable] and rely on existing deserializable types.
+/// You should use a visitor only when you deserialize a map or a union of several types.
+///
+/// ### Examples
```
```suggestion
/// ## Examples
```

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+///         let mut name = None;
+///         for (key, value) in members {
+///             let key_range = key.range();
+///             let Some(key) = TokenText::deserialize(key, diagnostics) else {
```
For simplicity, can we use `String::deserialize` instead? It's an example, and using a `TokenText` without an explanation can cause confusion.

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+///     }
+/// }
+///
+/// enum Union {
+///     Bool(bool),
+///     Str(String),
+/// }
```
This example is complicated to understand because it doesn't show how's the original data format. I think we should mention the data format, so we can explain why the union use it. Also, we should explain how the visitor works when we have a union, I mean under the hood. There's a lot of magic going on.

Conaclos replied:
Do you think we can use JSON? or should I introduce a fake data format?

ematipico replied:
I think JSON is a good example. Maybe we can have more examples in the `json` module, and we can redirect users to that crate to have more "real world" examples.

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+    ) -> Option<Self::Output> {
+        debug_assert!(
+            !Self::EXPECTED_TYPE.contains(ExpectedType::NULL),
+            "This method should be implemented because the type is expected."
```
```suggestion
            "This method should be implemented because the expected type is `null`."
```
I can see that we repeated the debug assert message. We should reword it, because "the type is expected" doesn't tell anything to me - as a consumer of this API. We can go beyond and tell the developer which type we expect in each function.

ematipico commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -1,31 +1,346 @@
+}
 
 /// A small type to interrogate the result of a JSON deserialization
 #[derive(Debug, Default)]
 pub struct Deserialized<P> {
     diagnostics: Vec<Error>,
-    deserialized: P,
+    deserialized: Option<P>,
 }
```
Before, `P` was always guaranteed; now it's not. We should add a comment to `deserialized` and explain at least when/why it's an `Option` now.

ematipico commented on crates/biome_formatter/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ version              = "0.2.0"
 
 [dependencies]
 biome_console     = { workspace = true }
+biome_deserialize = { workspace = true }
```
Where do we use this crate?

Conaclos replied:
It is used to implement LineWidth deserializer

ematipico replied:
Ah shoot sorry, I read `biome_console` 🤣

Conaclos replied:
Oh maybe you though about [biome_js_formatter's Crago.toml](https://github.com/biomejs/biome/pull/683/files#diff-02ab6043dc93f21e4b8e944e3657bad672e6303e7ac00fd67a5bb0975ad66317)?

ematipico commented on crates/biome_project/tests/invalid/lisence_not_string.json.snap:
```diff
@@ -4,7 +4,7 @@ expression: lisence_not_string.json
 lisence_not_string.json:2:13 deserialize ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 
-  × The value of key license is incorrect. Expected a string.
+  × Incorrect type, expected a string.
```
It's a bit sad that the message is generic :(

Conaclos replied:
Yeah. I am not sure how to fix it. I have to think about that.

Conaclos replied:
Yeah. I am not sure how to fix it without making the API more complex. I have to think about that.

ematipico replied:
I think it's fine for now, I understand the drawbacks of the refactor

Conaclos commented on crates/biome_deserialize/src/diagnostics.rs:
```diff
@@ -5,8 +5,47 @@ use biome_diagnostics::{
+            return write!(fmt, "no value");
+        }
+        let mut is_not_first = false;
+        for expected_type in self.iter() {
```
The final string?

ematipico replied:
Yes. If we expect multiple types, the error message should have something like this `Expected string, or number, or boolean but received ... `

We should make sure that the message is correctly formatted.

Conaclos replied:
I think I should implement `std::fmt::Display` instead of `biome_console::Display`. In the same way I should certainly use `impl std::fmt::Display` in all `DeserialziationDiagnostic` function? Not sure. A type that implements `std::fmt::Display` should implement `biome_console::Display`?

Conaclos replied:
I find a workaround by using `format_args!`.

Conaclos commented on crates/biome_deserialize/src/json.rs:
```diff
@@ -582,34 +99,82 @@ where
+        visitor: V,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<V::Output> {
+        let range = AstNode::range(&self);
```
Unfortunately I cannot make this change, because it is ambiguous for the compiler.

ematipico replied:
Really?? So, what's the point of the function `fn range()` here? 🤔 Sorry I need to understand 😓

Conaclos replied:
It is for retrieving the range of a desrializable value (which is under the hood an AST Node). Thus, we have two methods named `range`: one is from the `AstNode` trait, while the other is from `DeserializableValue`. `DeserializableValue::range` is calling `AstNode::range`.

Using `DeserializableValue` instead of `AstNode<L>` improves API ergonomics (no more generics to specify) and hide low-level details.

An alternative could be of introducing a trait `Ranged` that exposes a method `range`. `AstNode` and `DeserializableValue` could require the implementation of this trait. This could avoid the duplication.

ematipico replied:
I understand. It was a weird thing to see, for sure 🤔

migueldaipre commented on crates/biome_project/tests/invalid/lisence_not_string.json.snap:
```diff
@@ -4,7 +4,7 @@ expression: lisence_not_string.json
 ---
 lisence_not_string.json:2:13 deserialize ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```
The file name has a typo (lisence_not_string -> license_not_string)

Conaclos commented:
@ematipico I think I addressed all your concerns., but one:

 I am still unsure what to do when a member of an array or a map has a parsing error or a deserialization error.

I see several possibilities:

1. ignore members that cannot be deserialized
2. ignore members that cannot be deserialized for unordered data structures (HashMap and HasSet) and make the deserialization failed for ordered data structures (Vec, IndexSet, IndexMap) if at least one member cannot be deserialized.
3. make deserilization failed if at least one member cannot be deserialized.
4. Make the behavior configurable by passing an option to every deseriliazer. This will affect deserialization in the subtree.

I think (1) is the best, or maybe (2) as a tradeoff.
I am not fond of (4) because it adds a fourth parameter to `Deserialize::deserialize`.
Otheriwe, we could add `Deserialize::deserialize_with_options`.

ematipico commented:
I also think the first option is the best!
